streams-tests/shared/src/test/scala/zio/stream/ZStreamSpec.scala

```diff
@@ -2804,18 +2804,30 @@ object ZStreamSpec extends ZIOBaseSpec {
             } yield assert(res)(equalTo(List(1, 2, 3, 4)))
           )
         ),
-        testM("repeatEffectWith")(
-          for {
-            ref <- Ref.make[List[Int]](Nil)
-            fiber <- ZStream
-                      .repeatEffectWith(ref.update(1 :: _), Schedule.spaced(10.millis))
-                      .take(2)
-                      .runDrain
-                      .fork
-            _      <- TestClock.adjust(50.millis)
-            _      <- fiber.join
-            result <- ref.get
-          } yield assert(result)(equalTo(List(1, 1)))
+        suite("repeatEffectWith")(
+          testM("succeed")(
+            for {
+              ref <- Ref.make[List[Int]](Nil)
+              fiber <- ZStream
+                        .repeatEffectWith(ref.update(1 :: _), Schedule.spaced(10.millis))
+                        .take(2)
+                        .runDrain
+                        .fork
+              _      <- TestClock.adjust(50.millis)
+              _      <- fiber.join
+              result <- ref.get
+            } yield assert(result)(equalTo(List(1, 1)))
+          ),
+          testM("allow schedule rely on effect value")(checkM(Gen.int(1, 100)) { (length: Int) =>
+            var x        = length
```

Comment:
If you want to countdown from `length` to `0` you can, for example, allocate a `Ref[Int]` and use `ref.getAndUpdate`.

Reply:
Right, should've thought about that